PWGJE/Core/JetBkgSubUtils.cxx

```diff
@@ -0,0 +1,182 @@
+      rhovector.push_back(ijet.perp() / ijet.area());
+      totaljetAreaPhys += ijet.area();
+    }
+    // Full area
+    totalAreaCovered += ijet.area();
+  }
+  // calculate Rho as the median of the jet pT / jet area
+  double rho = TMath::Median<double>(rhovector.size(), rhovector.data());
+
+  if (doAreaSparse) {
+    // calculate The ocupancy factor, which the ratio of covered area / total area
+    double occupancyFactor = totalAreaCovered > 0 ? totaljetAreaPhys / totalAreaCovered : 1.;
+    rho *= occupancyFactor;
+  }
+
+  return rho;
+}
+
+double JetBkgSubUtils::estimateRhoPerpCone(const std::vector<fastjet::PseudoJet>& inputParticles, const std::vector<fastjet::PseudoJet>& jets)
+{
+
+  if (inputParticles.size() == 0 || jets.size() == 0) {
+    return 0.0;
+  }
+
+  double perpPtDensity1 = 0;
+  double perpPtDensity2 = 0;
+
+  fastjet::Selector selLeadingJet = fastjet::SelectorEtaRange(mBkgEtaMin, mBkgEtaMax) && fastjet::SelectorPhiRange(mBkgPhiMin, mBkgPhiMax);
+
+  std::vector<fastjet::PseudoJet> leadingJets = fastjet::sorted_by_pt(selLeadingJet(jets));
+
+  if (leadingJets.size() == 0) {
+    return 0.0;
+  }
+
+  fastjet::PseudoJet leadingJet = leadingJets[0];
+
+  double dPhi1 = 999.;
+  double dPhi2 = 999.;
+  double dEta = 999.;
+  double PerpendicularConeAxisPhi1 = 999., PerpendicularConeAxisPhi2 = 999.;
+  // build 2 perp cones in phi around the leading jet (right and left of the jet)
+  PerpendicularConeAxisPhi1 = ((leadingJet.phi() + (M_PI / 2.)) > (2 * M_PI)) ? leadingJet.phi() - ((3. / 2.) * M_PI) : leadingJet.phi() + (M_PI / 2.);
+  PerpendicularConeAxisPhi2 = ((leadingJet.phi() - (M_PI / 2.)) < 0) ? leadingJet.phi() + ((3. / 2.) * M_PI) : leadingJet.phi() - (M_PI / 2.);
+
+  for (auto& particle : inputParticles) {
+    // sum the momentum of all paricles that fill the two cones
+    dPhi1 = TMath::Abs(particle.phi() - PerpendicularConeAxisPhi1);
+    dPhi1 = (dPhi1 > M_PI) ? 2 * M_PI - dPhi1 : dPhi1;
+    dPhi2 = TMath::Abs(particle.phi() - PerpendicularConeAxisPhi2);
+    dPhi2 = (dPhi2 > M_PI) ? 2 * M_PI - dPhi2 : dPhi2;
+    dEta = leadingJet.eta() - particle.eta();
+    if (TMath::Sqrt(dPhi1 * dPhi1 + dEta * dEta) <= mJetBkgR) {
+      perpPtDensity1 += particle.perp();
+    }
+
+    if (TMath::Sqrt(dPhi2 * dPhi2 + dEta * dEta) <= mJetBkgR) {
+      perpPtDensity2 += particle.perp();
+    }
+  }
+
+  // Caculate rho as the ratio of average pT of the two cones / the cone area
+  Double_t perpPtDensity = (perpPtDensity1 + perpPtDensity2) / (2 * M_PI * mJetBkgR * mJetBkgR);
+
+  return perpPtDensity;
+}
+
+/// Sets the background subtraction estimater pointer
+void JetBkgSubUtils::setBkgE()
+{
+  bkgE = decltype(bkgE)(new fastjet::JetMedianBackgroundEstimator(mSelRho, mJetDefBkg, mAreaDefBkg));
+}
+
+/// Sets the background subtraction object
+fastjet::Subtractor JetBkgSubUtils::setSub(std::vector<fastjet::PseudoJet>& inputParticles, float& rhoParam, BkgSubMode bkgSubMode, const std::vector<fastjet::PseudoJet>& jets)
```

Comment:
you can put the jet vector after the particle one in case we want to set default values for the other quantities later

Reply:
rhoParam is given by reference so it can't have a default, but jets vector is needed only in the perp cone method so it needs to be empty by default.